### PR TITLE
PR: Hide tour tips dialog when Spyder looses focus

### LINKED ISSUE #3461: Click on outisde of tour should exit tour

### PR DESCRIPTION
**Pull request by mariacamilaremolinagutierrez**
Fixes #3461

### CODE REVIEW

File: spyder/app/tour.py [hunk 1]
```diff
@@ -524,14 +527,29 @@ def mousePressEvent(self, event):
         """Override Qt method"""
         pass
 
+    def focusInEvent(self, event):
+        """Reimplement Qt method."""
+
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Remove this blank.

File: spyder/app/tour.py [hunk 2]
```diff
@@ -524,14 +527,29 @@ def mousePressEvent(self, event):
...
+    def focusInEvent(self, event):
+        """Reimplement Qt method."""
+
+        # To be used so tips do not appear outside spyder
+        if self.hasFocus():
+            self.tour.tour_gain_focus()
+
+    def focusOutEvent(self, event):
+        """Reimplement Qt method."""
+
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Remove this blank.

File: spyder/app/tour.py [hunk 3]
```diff
@@ -815,6 +834,12 @@ def mousePressEvent(self, event):
 #            if clicked_widget == self.label_current:
 #            self.context_menu_requested(event)
 
+    def focusOutEvent(self, event):
+        """Reimplement Qt method."""
+
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Remove blank

File: spyder/app/tour.py [hunk 4]
```diff
@@ -987,6 +1018,7 @@ def _process_widgets(self, names, spy_window):
 
     def _set_data(self):
         """ """
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Add missing docstring

File: spyder/app/tour.py [hunk 5]
```diff
@@ -1174,6 +1212,19 @@ def close_tour(self):
         except:
             pass
 
+        self.is_tour_running = False
+
+    def hide_tips(self):
+        """ """
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Add docstring

File: spyder/app/tour.py [hunk 6]
```diff
@@ -1174,6 +1212,19 @@ def close_tour(self):
...
 
+        self.is_tour_running = False
+
+    def hide_tips(self):
+        """ """
+        self._clear_canvas()
+        self.tips.fade_out(self._hiding)
+
+    def unhide_tips(self):
+        """ """
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Add docstring

File: spyder/app/tour.py [hunk 7]
```diff
@@ -1200,6 +1251,22 @@ def first_step(self):
         """ """
         self.go_to_step(0)
 
+    def tour_lost_focus(self):
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Add docstring

File: spyder/app/tour.py [hunk 8]
```diff
@@ -1200,6 +1251,22 @@ def first_step(self):
         """ """
         self.go_to_step(0)
 
+    def tour_lost_focus(self):
+        if (self.is_tour_running and not self.any_has_focus() and
+            not self.setting_data and not self.tour_hidden):
+            self.hide_tips()
+
+    def tour_gain_focus(self):
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Add docstring

File: spyder/app/tour.py (hunk 4 above)
Comment (mariacamilaremolinagutierrez) [inline spyder/app/tour.py]: I didn't write this function, but I can write it though, I just want to be sure. Should I?

Comment (ccordoba12) [inline spyder/app/tour.py]: Don't worry, it's not a problem if you didn't write it :-)

File: spyder/app/tour.py [hunk 9]
```diff
@@ -898,6 +922,10 @@ def __init__(self, parent):
         # To capture the arrow keys that allow moving the tour
         self.tips.sig_key_pressed.connect(self._key_pressed)
 
+        # To control the focus of tour
+        self.setting_data = False
+        self.tour_hidden = False
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Change this to `hidden`

File: spyder/app/tour.py [hunk 10]
```diff
@@ -1154,26 +1188,40 @@ def start_tour(self):
...
         # Reset step to begining
         self.step_current = self.last_frame_active
-
         # Adjust the canvas size to match the main window size
         self.canvas.setFixedSize(width, height)
         self.canvas.move(QPoint(x, y))
         self.canvas.fade_in(self._move_step)
         self._clear_canvas()
 
+        self.is_tour_running = True
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Change this to `is_running`

File: spyder/app/tour.py [hunk 11]
```diff
@@ -1200,6 +1248,24 @@ def first_step(self):
         """ """
         self.go_to_step(0)
 
+    def tour_lost_focus(self):
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Change this to `lost_focus`

File: spyder/app/tour.py [hunk 12]
```diff
@@ -1200,6 +1248,24 @@ def first_step(self):
         """ """
         self.go_to_step(0)
 
+    def tour_lost_focus(self):
+        """Confirm if the tour loses focus and hides the tips."""
+        if (self.is_tour_running and not self.any_has_focus() and
+            not self.setting_data and not self.tour_hidden):
+            self.hide_tips()
+
+    def tour_gain_focus(self):
```
Comment (ccordoba12) [inline spyder/app/tour.py]: Change to `gain_focus`

**Comment by ccordoba12** [comment]
Thanks @mariacamilaremolinagutierrez for working on this one :-)

It looks good to me except for several new methods and attributes of the `AnimatedTour` class, that you prefixed with `tour_`, which I consider really unnecessary ;-)

**Comment by mariacamilaremolinagutierrez** [comment]
Thank you @ccordoba12 this is corrected now :)

File: spyder/app/tour.py [hunk 13]
```diff
@@ -815,6 +832,11 @@ def mousePressEvent(self, event):
 #            if clicked_widget == self.label_current:
 #            self.context_menu_requested(event)
 
+    def focusOutEvent(self, event):
+        """Reimplement Qt method."""
```
Comment (goanpeca) [inline spyder/app/tour.py]: This is probably a super minor comment.

But the convention we use is 

`Override Qt method.`

Comment (ccordoba12) [inline spyder/app/tour.py]: Ok, no prob. @mariacamilaremolinagutierrez, please change *Reimplement* for *Override* in all docstrings you introduced :-)

**Comment by goanpeca** [comment]
@ccordoba12 yes, its loooking great :-p, thanks @mariacamilaremolinagutierrez for digging into this pit of snakes!

**Comment by ccordoba12** [comment]
This looks great, thanks a lot @mariacamilaremolinagutierrez!!

**Pull request merged**